primitives/forest-verifier/src/lib.rs

```diff
@@ -0,0 +1,212 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::sp_runtime::DispatchError;
+use sp_std::{collections::btree_set::BTreeSet, vec::Vec};
+use sp_trie::{CompactProof, TrieDBBuilder, TrieLayout};
+use storage_hub_traits::CommitmentVerifier;
+use trie_db::TrieIterator;
+
+#[cfg(test)]
+mod tests;
+
+/// A struct that implements the `CommitmentVerifier` trait, where the commitment
+/// is a Merkle Patricia Trie root hash and the response to a challenge is given
+/// by either the exact key or the next and previous keys in the trie.
+pub struct ForestVerifier<T: TrieLayout, const H_LENGTH: usize>
+where
+    <T::Hash as sp_core::Hasher>::Out: for<'a> TryFrom<&'a [u8; H_LENGTH]>,
+{
+    pub _phantom: core::marker::PhantomData<T>,
```

Comment:
nipick: normally when there are no fields in a struct, you can have a tuple instead - `ForestVerifier(core::marker::PhantomData<T>)`.